src/crimson/os/seastore/seastore_types.h

```diff
@@ -1005,41 +1010,541 @@ constexpr journal_seq_t JOURNAL_SEQ_MAX{
 // JOURNAL_SEQ_NULL == JOURNAL_SEQ_MAX == journal_seq_t{}
 constexpr journal_seq_t JOURNAL_SEQ_NULL = JOURNAL_SEQ_MAX;
 
-// logical addr, see LBAManager, TransactionManager
-using laddr_t = uint64_t;
-constexpr laddr_t L_ADDR_MIN = std::numeric_limits<laddr_t>::min();
-constexpr laddr_t L_ADDR_MAX = std::numeric_limits<laddr_t>::max();
-constexpr laddr_t L_ADDR_NULL = L_ADDR_MAX;
-constexpr laddr_t L_ADDR_ROOT = L_ADDR_MAX - 1;
-constexpr laddr_t L_ADDR_LBAT = L_ADDR_MAX - 2;
+// logical offset, see LBAManager, TransactionManager
+using extent_len_t = uint32_t;
+constexpr extent_len_t EXTENT_LEN_MAX =
+  std::numeric_limits<extent_len_t>::max();
+
+using extent_len_le_t = ceph_le32;
+inline extent_len_le_t init_extent_len_le(extent_len_t len) {
+  return ceph_le32(len);
+}
+
+namespace details {
+// some concept required by laddr_t
+
+template <typename L, typename R = L>
+concept is_three_way_comparable = requires(const L &l, const R &r) { l <=> r; };
+
+template <typename T>
+concept is_convertible_to_ostream = std::is_convertible_v<T, std::ostream &>;
+
+template <typename T>
+concept is_streamable = requires(std::ostream &os, T value) {
+  { os << value } -> is_convertible_to_ostream;
+};
+
+} // namespace details
+
+/**
+ * laddr_t
+ *
+ * laddr_t uses uint128_t to represent the logical address internally.
+ * The layout of laddr:
+ *
+ * [pool:8][shard:8][crush:32][random:31][metadata:1][recover:1][local_clone_id:32][offset:15]
```

Comment:
As discussed, let's move the static flags (assume there are 4 bits) to the most significant bits.